File: frontend/services/permisos.ts
/**
 * ESPEJO de backend/utils/permisos.py. Fuente canónica = backend.
 * Mantener sincronizado. (Deuda: GET /api/auth/me eliminaría esta duplicación — Entrega 3.)
 *
 * Solo UX: filtra/oculta lo que el rol no puede usar. El control de seguridad real es
 * el backend (responde 403). Fail-closed: rol desconocido/null → sin acceso.
 */
import { getSession } from "@/services/api"
import type { UserRol } from "@/types/auth"

export type Accion = "read" | "write"

export type Seccion =
  | "empleados" | "areas" | "ausencias" | "vacaciones" | "vacantes"
  | "candidatos" | "onboarding" | "offboarding" | "costos" | "sucesion"
  | "assessment" | "organigrama" | "dashboard" | "empresa" | "reportes"
  | "importacion" | "integraciones" | "capacitaciones" | "evaluaciones"
  | "inventario" | "objetivos" | "usuarios" | "procesos" | "proyectos"

const MANDOS_MEDIOS_SECCIONES: ReadonlySet<Seccion> = new Set<Seccion>([
  "vacaciones",
  "ausencias",
])

/** Espejo de puede() del backend. 3 ramas + fail-closed ante rol desconocido/null. */
export function puede(rol: UserRol | null, seccion: Seccion, accion: Accion): boolean {
  if (rol === "admin_rrhh") return true
  if (rol === "gerencia_lectura") return accion === "read"
  if (rol === "mandos_medios") return MANDOS_MEDIOS_SECCIONES.has(seccion)
  return false
}

/** Rol del usuario logueado, o null si no hay sesión. */
export function getRol(): UserRol | null {
  return getSession()?.user.rol ?? null
}

/**
 * Mapa del primer segmento de la ruta → Seccion. Las rutas no listadas
 * (p. ej. /dashboard, /configuracion) devuelven null = siempre accesibles, no gateadas.
 */
const RUTA_SECCION: Readonly<Record<string, Seccion>> = {
  empleados: "empleados",
  areas: "areas",
  ausencias: "ausencias",
  vacaciones: "vacaciones",
  vacantes: "vacantes",
  onboarding: "onboarding",
  offboarding: "offboarding",
  costos: "costos",
  sucesion: "sucesion",
  assessment: "assessment",
  organigrama: "organigrama",
  empresas: "empresa",
  reportes: "reportes",
  capacitaciones: "capacitaciones",
  evaluaciones: "evaluaciones",
  inventario: "inventario",
  objetivos: "objetivos",
  procesos: "procesos",
  proyectos: "proyectos",
}

/** Seccion correspondiente a un pathname, o null si la ruta no se gatea por permiso. */
export function seccionDeRuta(pathname: string): Seccion | null {
  const seg = pathname.split("/").filter(Boolean)[0]
  if (!seg) return null
  return RUTA_SECCION[seg] ?? null
}

/**
 * Rutas gateables en el orden del sidebar, con su sección. Sirve para elegir el
 * destino de redirect cuando el rol no puede ver la ruta actual: se toma la primera
 * que sí puede leer (dashboard incluido, que admin y gerencia siempre leen).
 */
const RUTAS_ORDENADAS: ReadonlyArray<{ ruta: string; seccion: Seccion }> = [
  { ruta: "/dashboard", seccion: "dashboard" },
  { ruta: "/procesos", seccion: "procesos" },
  { ruta: "/proyectos", seccion: "proyectos" },
  { ruta: "/empresas", seccion: "empresa" },
  { ruta: "/empleados", seccion: "empleados" },
  { ruta: "/organigrama", seccion: "organigrama" },
  { ruta: "/vacantes", seccion: "vacantes" },
  { ruta: "/vacaciones", seccion: "vacaciones" },
  { ruta: "/ausencias", seccion: "ausencias" },
  { ruta: "/onboarding", seccion: "onboarding" },
  { ruta: "/offboarding", seccion: "offboarding" },
  { ruta: "/costos", seccion: "costos" },
  { ruta: "/sucesion", seccion: "sucesion" },
  { ruta: "/capacitaciones", seccion: "capacitaciones" },
  { ruta: "/evaluaciones", seccion: "evaluaciones" },
  { ruta: "/inventario", seccion: "inventario" },
  { ruta: "/objetivos", seccion: "objetivos" },
  { ruta: "/reportes", seccion: "reportes" },
]

/** Primera ruta (en orden de nav) que el rol puede leer, o null si ninguna (fail-closed). */
export function primeraRutaPermitida(rol: UserRol | null): string | null {
  const item = RUTAS_ORDENADAS.find((r) => puede(rol, r.seccion, "read"))
  return item ? item.ruta : null
}

File: backend/utils/permisos.py
"""
Núcleo de permisos funcionales (Entrega 2).

Define el modelo de capacidades por rol y la dependency factory que lo aplica
en los routers. Es deliberadamente AUTOCONTENIDO: no importa settings, supabase
ni anthropic, así que no ejecuta IO en import-time y puede testearse como función
pura sin DB ni HTTP.

Modelo de roles:
    admin_rrhh        → acceso total (lectura + escritura) en toda sección.
    gerencia_lectura  → solo lectura en toda sección.
    mandos_medios     → lectura y escritura solo en vacaciones y ausencias.

El enforcement es por dependency (Depends(require_permission(...))), nunca por
middleware. El cableado a cada router se hace en sub-tareas posteriores (16.3/16.4);
acá solo vive el núcleo.
"""
from enum import Enum
from typing import Awaitable, Callable, Optional, Union

from starlette.requests import Request

from utils.errors import AppError


class Accion(str, Enum):
    """Tipo de operación sobre una sección: lectura o escritura."""

    READ = "read"
    WRITE = "write"


class Seccion(str, Enum):
    """
    Conjunto cerrado de secciones del sistema. Una por módulo con router real
    registrado en main.py (auth queda fuera: no es una sección de negocio gateada).
    """

    EMPLEADOS = "empleados"
    AREAS = "areas"
    AUSENCIAS = "ausencias"
    VACACIONES = "vacaciones"
    VACANTES = "vacantes"
    CANDIDATOS = "candidatos"
    ONBOARDING = "onboarding"
    OFFBOARDING = "offboarding"
    COSTOS = "costos"
    SUCESION = "sucesion"
    ASSESSMENT = "assessment"
    ORGANIGRAMA = "organigrama"
    DASHBOARD = "dashboard"
    EMPRESA = "empresa"
    REPORTES = "reportes"
    IMPORTACION = "importacion"
    INTEGRACIONES = "integraciones"
    CAPACITACIONES = "capacitaciones"
    EVALUACIONES = "evaluaciones"
    INVENTARIO = "inventario"
    OBJETIVOS = "objetivos"
    USUARIOS = "usuarios"
    PROCESOS = "procesos"
    PROYECTOS = "proyectos"
    AUDITORIA = "auditoria"
    PERIODOS = "periodos"


# mandos_medios solo opera (R+W) sobre estas secciones; en el resto no puede nada.
MANDOS_MEDIOS_SECCIONES = frozenset({Seccion.VACACIONES, Seccion.AUSENCIAS})


def puede(
    rol: Optional[str],
    seccion: Union[Seccion, str],
    accion: Union[Accion, str],
) -> bool:
    """
    Decide si un rol puede ejecutar una acción sobre una sección. Función pura.

    Acepta tanto enums como strings en `seccion` y `accion`; los normaliza
    internamente. Es fail-closed: ante cualquier entrada inválida (rol None o
    desconocido, sección o acción que no existen en sus enums) retorna False.

    Args:
        rol: Rol del usuario ('admin_rrhh' | 'gerencia_lectura' | 'mandos_medios').
        seccion: Sección objetivo, como Seccion o su valor string.
        accion: Operación a realizar, como Accion o su valor string ('read'|'write').

    Returns:
        True si el rol tiene la capacidad pedida; False en cualquier otro caso.
    """
    try:
        seccion = Seccion(seccion)
        accion = Accion(accion)
    except (ValueError, TypeError):
        return False

    if rol == "admin_rrhh":
        return True
    if rol == "gerencia_lectura":
        return accion is Accion.READ
    if rol == "mandos_medios":
        return seccion in MANDOS_MEDIOS_SECCIONES
    return False


def require_permission(
    seccion: Union[Seccion, str],
    accion: Union[Accion, str] = Accion.READ,
) -> Callable[[Request], Awaitable[None]]:
    """
    Construye una dependency de FastAPI que exige permiso sobre (seccion, accion).

    El callable devuelto lee el rol desde request.state.user (seteado por
    AuthMiddleware), aplica puede() y lanza AppError FORBIDDEN (403) si no alcanza.
    Es fail-closed: si request.state.user no existe o no trae rol, deniega en vez
    de romper con AttributeError.

    Args:
        seccion: Sección que protege la ruta.
        accion: Operación requerida; READ por defecto.

    Returns:
        Dependency async para usar con Depends(require_permission(...)).
    """

    async def _verificar(request: Request) -> None:
        user = getattr(request.state, "user", None)
        rol = user.get("rol") if isinstance(user, dict) else None
        if not puede(rol, seccion, accion):
            raise AppError(
                "No tenés permiso para realizar esta acción", "FORBIDDEN", 403
            )

    return _verificar
